crates/core/tedge/src/cli/certificate/cli.rs

```diff
@@ -193,16 +196,187 @@ pub enum UploadCertCli {
     },
 }
 
-fn get_device_id_from_config(
+/// Returns the device ID from the config if no ID is provided by CLI
+/// If the provided device ID mismatches the one from the config, it returns an error
+fn get_device_id(
+    id: Option<String>,
     config: &TEdgeConfig,
+    config_location: &TEdgeConfigLocation,
     cloud: &Option<Cloud>,
-) -> Result<String, ReadError> {
-    let id = match cloud {
-        None => config.device.id(),
-        Some(Cloud::C8y(profile)) => config.c8y.try_get(profile.as_deref())?.device.id(),
-        Some(Cloud::Azure(profile)) => config.az.try_get(profile.as_deref())?.device.id(),
-        Some(Cloud::Aws(profile)) => config.aws.try_get(profile.as_deref())?.device.id(),
-    }?
-    .to_owned();
-    Ok(id)
+) -> Result<String, TEdgeError> {
+    let config_id = config.device_id(cloud.as_ref()).ok();
+    let writable_key = get_writable_key(cloud)?;
+    let dto = config_location
+        .load_dto_from_toml_and_env()
+        .context("failed to read tedge.toml file")?;
+
+    match id {
+        None => match config_id {
+            None => Err(anyhow!("No device ID is provided. Use `--device-id <name>` option to specify the device ID.").into()),
+            Some(config_id) => Ok(config_id.into()),
+        }
+        Some(input_id) if config_id.is_none() => Ok(input_id),
+        Some(input_id) if input_id == config_id.unwrap() => Ok(input_id),
+        Some(input_id) => match cloud.as_ref().map(<_>::into) {
+            None => Err(MismatchedDeviceId {
+                input_id,
+                config_id: config_id.unwrap().into(),
+                writable_key,
+            }),
+            Some(tedge_config::Cloud::C8y(profile)) => {
+                let key = profile.map(|name| name.to_string());
+                let c8y_dto = dto.c8y.try_get(key.as_deref(), "c8y")?;
+                if c8y_dto.device.id.is_some() {
+                    Err(MismatchedDeviceId {
+                        input_id,
+                        config_id: config_id.unwrap().into(),
+                        writable_key,
+                    })
+                } else {
+                    Ok(input_id)
+                }
+            }
+            Some(tedge_config::Cloud::Az(profile)) => {
+                let key = profile.map(|name| name.to_string());
+                let az_dto = dto.az.try_get(key.as_deref(), "az")?;
+                if az_dto.device.id.is_some() {
+                    Err(MismatchedDeviceId {
+                        input_id,
+                        config_id: config_id.unwrap().into(),
+                        writable_key,
+                    })
+                } else {
+                    Ok(input_id)
+                }
```

Comment:
I don't understand what is checked here (for `az` as well as for `c8y` and `aws`).

I seems that we have 3 device ids:  `input_id`, `config_id`, and `az_dto.device.id`, the value of the later being ignored.

Reply:
`dto` is used only to check if the value exists really in `az.device.id`. This cannot be checked by using `reader`.

Some examples:
---

```toml
[device]
id = "generic"

[az.device]
id = "az_id"
```

- `config_id` (= the value from `reader`) has `az_id`.
- `az.dto.device.id` has `Some("az_id")`. So, here the check `az_dto.device.id.is_some()` is `true`.

---

```toml
[device]
id = "generic"
```

- `config_id` (= the value from `reader`) has `generic`.
- - `az.dto.device.id` has `None`. So, here the check `az_dto.device.id.is_some()` is `false`.

Reply:
Okay, but then why don't we have a check of the cloud specific value. It might be set to the same as for the global device id. For example:

```
[device]
id = "az_id"

[az.device]
id = "az_id"
```

Reply:
@didier-wenzek 
I didn't get which error occurs in your case. In the case, using the value from Reader and Dto is eventually the same.

| key   | Reader |  Dto|
| -------- | ------- | -----|
|device.id  | az_id  | Some("az_id")
| az.device.id | az_id  | Some("az_id")

Reply:
> I didn't get which error occurs in your case. 

This is my point, there should be no error. While, if I understand correctly, the current code is returning an error because `az_dto.device.id.is_some()`.

Reply:
There is one match condition before entering this arm,
```
Some(input_id) if input_id == config_id.unwrap() => Ok(input_id),
```
So, if `input_id` is "az_id", the command will return `Ok(input_id)`.

And, yes, it returns an error if `input_id` != `config_id`and `az_dto.device_id.is_some()`. This should be an error because user tries to create a cert with mismatched ids.

Reply:
So, we are back to my first comment: I don't understand what is checked here.

I will have another look when https://github.com/thin-edge/thin-edge.io/pull/3326#discussion_r1922383288 will be addressed.

Reply:
Simplified in 99620aa

Also see the thread in https://github.com/thin-edge/thin-edge.io/pull/3326#discussion_r1923282326